### DISCUSSION TITLE
Rendering components within open text fields

### DISCUSSION THREAD

**Discussion by seancdavis** [ample/gatsby-starter-ample#194] [Ideas]
There are two parts to this discussion ...

## How to Render Components from Open Text Fields

The first part is about how we accomplish this. How can we achieve enabling content editors to have (even) more control by adding components within CMS fields that were originally meant just as open copy type of fields?

Should we look into supporting MDX for these fields? Or should it be more like a shortcode type of feature? Are there any services to help with either?

And does the approach change from Forestry to Contentful — i.e. can we maintain the same level of support regardless of the CMS we use?

## _Should_ we ...

Part two is less about _can_ and _how_ and more about if we _should_ incorporate this type of approach? Or should we keep it behind the curtain, since we can be so flexible and quick in building templates and components?

**Comment by dholloran-ample** [comment]
## How to Render Components from Open Text Fields

### Thoughts

I personally would try MDX first to see if it works. Since MDX in the content editor is no different than a Shortcode in WordPress. The only benefit from having something custom aka Shortcode would be we could use `[`/`]` instead of `<`/`>` to note that it's a special element (More a makes my brain happy thing I think).

### How to (possibly)
With Gatsby, MDX would probably be possible with the code below. The one thing I'm not sure of is if `contentFromCMS` can just be a string or if we need to run it through something to change it in some way. I want to say Gatsby MDX does something like this when it reads in the MDX files.

```
{/* MDX requires you to configure the components that can be used. */}
<MDXProvider components={allowedComponents}>
  {/* I believe passing the content straight from the CMS would work here in theory. */}
  <MDXRenderer>{contentFromCMS}</MDXRenderer>
</MDXProvider>;
```

### Notes
- I believe we went down this road and got fairly close in LSI to having a shortcode parser but it was semi flaky if I recall. So MDX (if it works) would mean one less thing we have to maintain but it could introduce weird build failures in the future. Currently, Gatsby MDX has some really cryptic errors mainly around whitespace, fragments, and comments.
- Contentful allows for embedding content types into at least the rich text fields. I do not remember if markdown is supported. This would not be an across the board solution unless Forestry has a comparable feature quick search does not show they do but I may have missed it.

## Should we...

> You were so preoccupied with whether or not you could. You didn't stop to think if you should.
> 
> Ian Malcolm - Jurassic Park

It's good that we are thinking about should we?

- Personally, I wouldn't for anything complex. Things like buttons, fancy backgrounds, or video embeds can be helpful. Anything that is widely re-usable and containable with a small API surface.
- I remember at one point thinking maybe shortcodes have been a good idea in WordPress. Which they are in theory until you get too many of them. Granted I've seen it taken to the extreme where whole sites have been built using it so I may be traumatized.
- I've also found content editors even if they are semi-technical can have problems with it without some sort of UI to fill in configuration values. At that point, you're almost better off wiring up the component to the CMS.

**Comment by seancdavis** [comment]
After talking this out separately, I agree with the approach to explore this and see how we could get there. I think we should start with only the following components:

- Button
- Image
- Video

A goal of this should be that it's trivial to support new components in this process. And what should come of this is (at least) a suggestion on how we're going to document this for our clients. The playgrounds are not built for this kind of documentation today. But we will likely want to figure out how we can provide a simple means of documenting usage to content editors.

---

Regarding implementation, my guess is that you're going to want to tap into the [gatsby-remark-ample](https://github.com/ample/gatsby-starter-ample/blob/main/plugins/gatsby-remark-ample/gatsby-node.js) process rather than using the gatsby mdx plugin. I say that because our remark plugin already handles a lot of customizations on top of the markdown files in the `src/content` directory. I worry that if we swap to MDX as the main source plugin that we're going to have to rework all those features.

I'd suggest getting familiar with how that plugin works before getting too far down into the exploration. I'd be happy to pair and/or provide some reasoning on why that plugin works the way it does. But a good starting point is [the plugin's README](https://github.com/ample/gatsby-starter-ample/blob/main/plugins/gatsby-remark-ample/README.md).

**Comment by seancdavis** [comment]
Also ...

![](https://media.giphy.com/media/mCClSS6xbi8us/giphy.gif)

**Comment by seancdavis** [comment]
Here's an article that may be of interest in this adventure: https://www.smashingmagazine.com/2020/12/mdx-stored-sanity-next-js-website/

It's written about Next, but being React, I'm guessing there's some insight to be gained.
